### DISCUSSION TITLE
sharing a database connection pool across (future) targets?

### DISCUSSION THREAD

**Discussion by mkoohafkan** [ropensci/targets#774] [Ideas]
As the title says, I a wondering if it is possible to share a database connection pool between targets, including targets spawned via `tar_make_future()`. My first thought was to try using the [`pool`]() package and defining an initial target with `tar_cue(mode = "always")`, but I question if that will actually work across `callr` sessions.

**Comment by wlandau** [comment]
Database connections are [usually not exportable](https://cran.r-project.org/web/packages/future/vignettes/future-4-non-exportable-objects.html). I recommend re-creating and tearing down the connection within each target that needs it. You can use [`tarchetypes::tar_hook_before()`](https://docs.ropensci.org/tarchetypes/reference/tar_hook_before.html) to insert code that creates a connection and assigns an `on.exit()` callback to close it.

**Reply by mkoohafkan** [reply]
Thanks, that's what I figured. Is there any advantage to using `tar_hook_before()` vs connecting (and disconnecting via `on.exit()` calls) to databases within the `tar_target()` expressions?

**Reply by wlandau** [reply]
They’re the same, tar_hook_before() is just a more compact way to write it for lots of targets.

**Reply by MxNl** [reply]
Should this also work with dynamic branching? Because I tried `tar_hook_before()` to create a pool object and close it `on.exit()` and it works for targets that have no branching but doesn't for branched targets. Here I get the error: Object POOL not found. Thanks!
